Fix pod loading when under restrictive RBAC

Nokel81 opened the pull request:
- Correctly handle the case where `undefined` is returned by the fetch because of a 403: Unauthorized being returned from the Kube Api due to only having list permissions in some of the namespaces

- Fix rendering of Pod Secrets in the details view to handle correctly the case where the user doesn't have list or get secrets permissions

- Remove CancelablePromise<T> as it was only used in one place and the return types where wrong causing a crash while debugging the above.

- Move HelmChartDetails to create its own AbortController for cancelling its request itself as that was the only instance of the cancel() method being called

- Significantly improve the typing of `isJsonApiData` and `isJsonApiDataList` so that they actually make sure that the form of the data matches the assertion they make. This also removes a crash from incorrectly assuming that `any` could not be `undefined`.

- Add tests for the above two functions

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixed #1961 , fixes #2227

Nokel81 committed:
Fix pod loading under some more RBAC restrictions

- Correctly handle the case where `undefined` is returned by the fetch
  because of a 403: Unauthorized being returned from the Kube Api due to
  only having list permissions in some of the namespaces

- Fix rendering of Pod Secrets in the details view to handle correctly
  the case where the user doesn't have list or get secrets permissions

- Remove CancelablePromise<T> as it was only used in one place and the
  return types where wrong causing a crash while debugging the above.

- Move HelmChartDetails to create its own AbortController for cancelling
  its request itself as that was the only instance of the cancel()
  method being called

- Significantly improve the typing of `isJsonApiData` and
  `isJsonApiDataList` so that they actually make sure that the form of
  the data matches the assertion they make. This also removes a crash
  from incorrectly assuming that `any` could not be `undefined`.

- Add tests for the above two functions

- Significantly improve the readability of the isKubeJson* functions

- add doc comments for `ItemStore.prototype.sortItems`

- add many more helper functions for type narrowing

- Add some more handling of error cases with RBAC

- Add notifications when errors occur (which leave lists in the loading
  state)

- properly set response codes when an error occurs for listing helm
  releases

- support kube API 1.20 with now optional selfLink

- fix KubeObjectStore.subscribe not waiting for the corisponding load to occur

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -346,27 +346,17 @@ export class Pod extends WorkloadKubeObject {
-        }
-      });
+      if (state.terminated) {
+        return state.terminated.reason || "Waiting";
```
Little bit of refactoring on the side?

```suggestion
        return state.terminated.reason || "Terminated";
```

jim-docker commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -477,10 +467,7 @@ export class Pod extends WorkloadKubeObject {
-    if (!this.spec.nodeSelector["kubernetes.io/os"] && !this.spec.nodeSelector["beta.kubernetes.io/os"]) return;
-
-    return this.spec.nodeSelector["kubernetes.io/os"] || this.spec.nodeSelector["beta.kubernetes.io/os"];
+    return this.spec.nodeSelector?.["kubernetes.io/os"] || this.spec.nodeSelector?.["beta.kubernetes.io/os"];
```
more side refactoring. Shouldn't it be:
```suggestion
    return this.spec?.nodeSelector?.["kubernetes.io/os"] || this.spec?.nodeSelector?.["beta.kubernetes.io/os"];
```

Nokel81 replied:
I think that this is fine to be in this PR since I am already touching the `pods.api.ts` file.

With regards to your comment. No, spec is not an optional field so there is no point conditionally checking it.

jakolehm replied:
Maybe add a test for this while touching this method? 🙂

Nokel81 replied:
Sounds good

jim-docker commented on src/renderer/api/kube-object.ts:
```diff
@@ -55,12 +56,233 @@ export class KubeObject implements ItemObject {
+
+    if (typeof object.kind !== "string"
+      || typeof object.apiVersion !== "string"
+      || !object.metadata || typeof object.metadata !== "object"
```
is testing `!object.metadata` necessary since you already checked it with `hasOwnProperties()` call? Or you're trying to catch if it has value `null` or `undefined` or 0 or false...

Nokel81 replied:
I am trying to catch `null` (and would have used `object !== null` but you and Stefan both said you don't like that. 

For historical reasons `typeof null` is `"object"` so we have to add this check here.

jim-docker replied:
It's `==` that I said I didn't like. `!==` is fine.

Nokel81 replied:
oh I see

jim-docker replied:
But `!object.metadata` is fine here, just wanted to make sure you needed to do the check.

jim-docker commented on src/renderer/api/kube-object.ts:
```diff
@@ -55,12 +56,233 @@ export class KubeObject implements ItemObject {
+    if (hasOwnProperty(metadata, "finalizers")
+      && typeof metadata.finalizers !== "undefined"
+      && !(Array.isArray(metadata.finalizers)
+      && metadata.finalizers.every(finalizer => typeof finalizer === "string"))
```
I almost missed that this is a compound conditional, should/could this line be indented to make it more obvious? Or something...

Nokel81 replied:
The `every` call?

jim-docker replied:
Something like this?
```
    if (hasOwnProperty(metadata, "finalizers")
      && typeof metadata.finalizers !== "undefined"
      && !(
        Array.isArray(metadata.finalizers)
        && metadata.finalizers.every(finalizer => typeof finalizer === "string")
      )
    ) {
 ```
Maybe young eyes don't need such assistance :)

Nokel81 replied:
Oh yes that would be better

jim-docker commented on src/renderer/api/kube-object.ts:
```diff
@@ -55,12 +56,233 @@ export class KubeObject implements ItemObject {
+            if (!hasOwnProperty(metadata.labels, label) || typeof metadata.labels[label] !== "string") {
+              return false;
+            }
+          }
```
Would `Object.entries(metadata.labels).forEach(...)` eliminate the need to use `hasOwnProperty()`?

Nokel81 replied:
This check technically verifies that there the values are not "up the prototype" chain. Furtherore, we cannot use `forEach` we would have to use `every` again because you cannot return out of several functions at once.

jim-docker replied:
I know, but it's 2021, we still have to rely on `hasOwnProperty`...

jim-docker commented on src/renderer/api/kube-object.ts:
```diff
@@ -55,12 +56,233 @@ export class KubeObject implements ItemObject {
+    return true;
+  }
+
+  static isPartialJsonApiData(object: unknown): object is Partial<KubeJsonApiData> {
```
I don't get the difference between this and `isJsonApiData()`. Can duplicated code be shared? They look the same except for what `object` is.

Nokel81 replied:
The difference is subtle yes, but then again so is the difference between `T` and `Partial<T>`. Namely, this function still returns true if any of the top level fields are `undefined` or not present. That is the only difference.

Unfortunately, there is no easy way to have a `isPartial` argument. However, I might be able to simplify the above by first calling this and then verifying that none of the fields are `undefined` or missing.

jim-docker commented on src/renderer/item.store.ts:
```diff
@@ -28,8 +28,8 @@ export abstract class ItemStore<T extends ItemObject = ItemObject> {
   @action
-  protected sortItems(items: T[] = this.items, sorting?: ((item: T) => any)[], order?: "asc" | "desc"): T[] {
-    return orderBy(items, sorting || this.defaultSorting, order);
+  protected sortItems(items: T[] = this.items, sorting: ((item: T) => any)[] = [this.defaultSorting], order?: "asc" | "desc"): T[] {
```
Either way this is unreadable to me. Modern JavaScript/Typescript is for the young.

Nokel81 replied:
I'll had some doc comments to help clear this up.

jim-docker reviewed (commented):
Nice quick work. Looks like strong bullet-proofing but touches some core code, need to test thoroughly for unexpected side-effects

jakolehm commented on src/renderer/components/+apps-helm-charts/helm-chart-details.tsx:
```diff
@@ -26,10 +25,11 @@ export class HelmChartDetails extends Component<Props> {
   @observable error: string = null;
 
-  private chartPromise: CancelablePromise<{ readme: string; versions: HelmChart[] }>;
+  private chartPromise: Promise<{ readme: string; versions: HelmChart[] }>;
```
Not sure how these changes are related to the PR title?

Nokel81 replied:
Specifically for this, see note in PR description

> Remove CancelablePromise as it was only used in one place and the return types where wrong causing a crash while debugging the above.

If you want I can move the out to a separate PR but I did encounter an exception while debugging this issue.

jakolehm replied:
Ok, missed the description. Probably fine to keep here as it's somewhat related.

jakolehm commented on extensions/example-extension/package-lock.json:
```diff
@@ -626,7 +626,644 @@
     "@k8slens/extensions": {
       "version": "file:../../src/extensions/npm/extensions",
-      "dev": true
+      "dev": true,
```
Please remove `extensions/*/package-lock.json` changes.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
 
   getAllContainers() {
-    return this.getContainers().concat(this.getInitContainers());
+    return [...this.getContainers(), ...this.getInitContainers()];
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
-    return this.getAllContainers().filter(container => {
-      return statuses.find(status => status.name === container.name && !!status.state["running"]);
-    }
+    const runningContainerNames = new Set(
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
-    if (containerStatuses) {
-      statuses.push(...containerStatuses);
-    }
+    const { containerStatuses = [], initContainerStatuses = [] } = this.status ?? {};
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
 
   getRestartsCount(): number {
-    const { containerStatuses } = this.status;
+    const { containerStatuses = [] } = this.status ?? {};
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
 
   getQosClass() {
-    return this.status.qosClass || "";
+    return this.status?.qosClass || "";
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
 
   getReason() {
-    return this.status.reason || "";
+    return this.status?.reason || "";
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -267,60 +267,55 @@ export class Pod extends WorkloadKubeObject {
-    const goodConditions = ["Initialized", "Ready"].every(condition =>
-      !!this.getConditions().find(item => item.type === condition && item.status === "True")
-    );
+    const trueConditionTypes = new Set(this.getConditions()
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -346,37 +341,27 @@ export class Pod extends WorkloadKubeObject {
 
   getStatusPhase() {
-    return this.status.phase;
+    return this.status?.phase;
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -346,37 +341,27 @@ export class Pod extends WorkloadKubeObject {
 
   getConditions() {
-    return this.status.conditions || [];
+    return this.status?.conditions || [];
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -406,20 +389,19 @@ export class Pod extends WorkloadKubeObject {
-    });
-    const crashLoop = !!this.getContainerStatuses().find(condition => {
-      const waiting = condition.state.waiting;
+    for (const { type, status } of this.getConditions()) {
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -473,14 +455,11 @@ export class Pod extends WorkloadKubeObject {
 
   getNodeName() {
-    return this.spec?.nodeName;
+    return this.spec.nodeName;
```
Why are we changing this? Please revert unless really related to the PR title.

jakolehm commented on src/renderer/api/endpoints/pods.api.ts:
```diff
@@ -473,14 +455,11 @@ export class Pod extends WorkloadKubeObject {
-
-    return this.spec.nodeSelector["kubernetes.io/os"] || this.spec.nodeSelector["beta.kubernetes.io/os"];
+  getSelectedNodeOs(): string | undefined {
+    return this.spec.nodeSelector?.["kubernetes.io/os"] || this.spec.nodeSelector?.["beta.kubernetes.io/os"];
```
Why are we changing this? Please revert unless really related to the PR title.

aleksfront commented on src/renderer/api/endpoints/helm-charts.api.ts:
```diff
@@ -31,11 +31,11 @@ export const helmChartsApi = {
   },
 
-  get(repo: string, name: string, readmeVersion?: string) {
+  get(repo: string, name: string, readmeVersion?: string, reqInit?: RequestInit) {
```
Currently we're not following any guidlines regarding to this, so nothing critical here, but I beleive it should be considered twice because of this quote from Robert C. Martin "Clean Code".

> Functions should have a small number of arguments. No arguments is best, followed by one, two, and three. More than three is very questionable and should be avoided with prejudice.

Nokel81 replied:
Maybe, the last two arguments because they are both optional.

Nokel81 replied:
I think that this work should be done in a seperate PR

aleksfront commented on src/renderer/api/json-api.ts:
```diff
@@ -119,48 +118,53 @@ export class JsonApi<D = JsonApiData, P extends JsonApiParams = JsonApiParams> {
       reqInit,
     };
 
-    return cancelableFetch(reqUrl, reqInit).then(res => {
-      return this.parseResponse<D>(res, infoLog);
-    });
```
Any specific reason we're changing promises to async/await functions?

Nokel81 replied:
Please see the comments that I made in the initial post for this PR. It explains why I have done these changes.

aleksfront commented on src/renderer/api/kube-api.ts:
```diff
@@ -302,55 +294,85 @@ export class KubeApi<T extends KubeObject = any> {
     }
 
-    return data;
+    return null;
```
Why we returning `null` here instead of `data`? Couldn't unprocessed `data` be used in some places?

Nokel81 replied:
It could but that is not the point of a validation function, I guess the other option would be the throw an error.

aleksfront commented on src/common/utils/type-narrowing.ts:
```diff
@@ -0,0 +1,89 @@
+ * checks if val is of type object and isn't null
+ * @param val the value to be checked
+ */
+export function isObject(val: unknown): val is object {
+  return typeof val === "object" && val !== null;
+}
```
Arrays will pass here since `typeof [1, 2] === object`.

Nokel81 replied:
I know, but generally you will want to do additional validation, even if it is just that all keys and values are strings (for `Record<string, string>`).

aleksfront commented on src/renderer/components/dock/log.store.ts:
```diff
@@ -65,72 +73,73 @@ export class LogStore {
+  }
 
   /**
    * Converts logs into a string array
-   * @returns {number} Length of log lines
+   * @returns Length of log lines
    */
```
Another signal that such comments should be avoided in overall. Once described they tempted to get away from code and started to provide lies. Moreover, this description was invalid right from the start.

The function code was changed, but description comment left untouched.

Nokel81 replied:
I *emphatically* do not agree that we should avoid comments because the code might be updated.

We should as a team require in code reviews that all comments should be updated or at least marked as checked whenever we change the functions. This codebase is slowly getting larger and we should leave helpful guides to future maintainers about what the code does.

aleksfront commented on src/renderer/components/dock/log.store.ts:
```diff
@@ -65,72 +73,73 @@ export class LogStore {
-
-    return logs ? logs.length : 0;
+  get lines(): number {
+    return this.logs.length;
```
Very nice! 👍

Nokel81 committed:
wrap all lifetime functions with a fail on async reject

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix abort-controller to use polyfill

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/kube-object.store.ts:
```diff
@@ -272,17 +279,21 @@ export abstract class KubeObjectStore<T extends KubeObject = any> extends ItemSt
-        });
-      });
-    }
+    // This waits for
```
what?

jim-docker commented on src/renderer/api/json-api.ts:
```diff
@@ -119,48 +117,53 @@ export class JsonApi<D = JsonApiData, P extends JsonApiParams = JsonApiParams> {
+    }
+
+    if (status >= 200 && status < 300) {
+      console.log(data, res);
```
is this for debugging?

jim-docker commented on src/renderer/api/json-api.ts:
```diff
@@ -119,48 +117,53 @@ export class JsonApi<D = JsonApiData, P extends JsonApiParams = JsonApiParams> {
+      this.onError.emit(error, res);
+      this.writeLog({ ...log, error });
+      throw error;
+    }
```
```suggestion
    }
    
    const error = new JsonApiErrorParsed(data, this.parseError(data, res));

    this.onError.emit(error, res);
    this.writeLog({ ...log, error });
    throw error;
```

jim-docker commented on src/renderer/api/kube-api.ts:
```diff
@@ -303,55 +300,85 @@ export class KubeApi<T extends KubeObject = any> {
+
+    const parsed = this.parseResponse(res);
+
+    if (Array.isArray(parsed)) {
```
Remind me, what does it mean if `parsed` is an array? I know `get()` should return a single item but what if the array has only one item?

jim-docker commented on src/renderer/components/+namespaces/add-namespace-dialog.tsx:
```diff
@@ -33,20 +33,18 @@ export class AddNamespaceDialog extends React.Component<Props> {
+      const created = await namespaceStore.create({ name: namespace });
+
+      onSuccess?.(created);
+      AddNamespaceDialog.close();
```
should this be in a `finally` block instead?

Nokel81 replied:
No, because the user might want to fix what caused the error to happen and try again.

jim-docker commented on src/renderer/kube-object.store.ts:
```diff
@@ -291,60 +302,47 @@ export abstract class KubeObjectStore<T extends KubeObject = any> extends ItemSt
-            });
-          }, 5000);
+          clearTimeout(timedRetry);
+          timedRetry = setTimeout(watch, 5000);
```
don't need this in `else` block

Nokel81 committed:
Hide UserManagement sidebar based on access control

- Also throw errors in kube-api methods if an array is returned by the
  API but a single item was expected or vice-versa

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/components/+user-management/user-management.tsx:
```diff
@@ -13,29 +13,36 @@ import { isAllowedResource } from "../../../common/rbac";
+      });
+    }
+
+    if (isAllowedResource("role")) {
```
what about "clusterrole"?

Nokel81 replied:
Ah thanks forgot about that one.

Nokel81 committed:
all Roles tab to be viewed if cluster has access to clusterrole

Signed-off-by: Sebastian Malton <sebastian@malton.name>
